### PR TITLE
Fix angle generation caching not working with irregular chunks

### LINKED ISSUE #1973: Using cached geolocation and angles results in an error if chunk size not appropriately set.
**Describe the bug**
The new cached geolocation and satellite angles fails to function correctly if the `PYTROLL_CHUNK_SIZE` is not set to an integer factor of the level 1 image size.

**To Reproduce**

```python
from satpy import Scene
from glob import glob
import satpy

satpy.config.set({'cache_sensor_angles': True})
satpy.config.set({'cache_lonlats': True})
satpy.config.set({'cache_dir': '/tmp/satpy/'})

files = glob(f'*.DAT')

comp = 'true_color'

scn = Scene(files, reader='ahi_hsd')
scn.load([comp])
scn2 = scn.resample(scn.coarsest_area(), resampler='native')
scn2.save_datasets()
```

**Expected behavior**
the output files to be created, along with cache zarr files for the angles/lats/lons if not already available.

**Actual results**
If `PYTROLL_CHUNK_SIZE` is not set to an integer factor of the AHI image size then I get the following error:
```
  File "/home/proud/PycharmProjects/satpy/satpy/scene.py", line 1260, in load
    self.generate_possible_composites(generate, unload)
  File "/home/proud/PycharmProjects/satpy/satpy/scene.py", line 1319, in generate_possible_composites
    keepables = self._generate_composites_from_loaded_datasets()
  File "/home/proud/PycharmProjects/satpy/satpy/scene.py", line 1340, in _generate_composites_from_loaded_datasets
    return self._generate_composites_nodes_from_loaded_datasets(needed_comp_nodes)
  File "/home/proud/PycharmProjects/satpy/satpy/scene.py", line 1346, in _generate_composites_nodes_from_loaded_datasets
    self._generate_composite(node, keepables)
  File "/home/proud/PycharmProjects/satpy/satpy/scene.py", line 1370, in _generate_composite
    prereq_datasets = self._get_prereq_datasets(
  File "/home/proud/PycharmProjects/satpy/satpy/scene.py", line 1452, in _get_prereq_datasets
    self._generate_composite(prereq_node, keepables)
  File "/home/proud/PycharmProjects/satpy/satpy/scene.py", line 1404, in _generate_composite
    composite = compositor(prereq_datasets,
  File "/home/proud/PycharmProjects/satpy/satpy/modifiers/geometry.py", line 72, in __call__
    coszen = get_cos_sza(vis)
  File "/home/proud/PycharmProjects/satpy/satpy/modifiers/angles.py", line 287, in get_cos_sza
    lons, lats = _get_valid_lonlats(data_arr.attrs["area"], data_arr.chunks)
  File "/home/proud/PycharmProjects/satpy/satpy/modifiers/angles.py", line 150, in __call__
    self._cache_results(res, zarr_format)
  File "/home/proud/PycharmProjects/satpy/satpy/modifiers/angles.py", line 178, in _cache_results
    new_sub_res = sub_res.to_zarr(zarr_path,
  File "/home/proud/miniconda3/lib/python3.9/site-packages/dask/array/core.py", line 2708, in to_zarr
    return to_zarr(self, *args, **kwargs)
  File "/home/proud/miniconda3/lib/python3.9/site-packages/dask/array/core.py", line 3427, in to_zarr
    raise ValueError(
ValueError: Attempt to save array to zarr with irregular chunking, please call `arr.rechunk(...)` first.
```

### PR DESCRIPTION
**Pull request by djhoese**
The Zarr format doesn't like irregular chunks, chunks that aren't all the same size in a single dimension except for the last chunk. This results in issues like #1973 where you can't use the caching unless you're very careful about the chunking of your data. This PR does a few important things:

1. It detects and normalizes any chunk request that is irregular and uses this regular chunking when calling the function being cached.
2. It saves the zarr array as this regular chunking, but then calls `.rechunk` with the original chunk size before giving the result back to the user.
3. The normalized chunks are used when the arguments are hashed which allows for different input chunks to use the same cached values.

 - [x] Closes #1973 
 - [x] Tests added 
 - [ ] Fully documented

### CODE REVIEW

**Comment by Plantain** [comment]
Works for me!

File: satpy/modifiers/angles.py [hunk 1]
```diff
@@ -183,6 +205,22 @@ def _cache_results(self, res, zarr_format):
...
+
+    It is assumed a tuple of tuples of integers is defining chunk sizes. If
+    a tuple like this is not found then arguments are checked for array-like
+    objects with a ``.chunks`` attribute.
+
+    """
+    chunked_args = [arg for arg in args if hasattr(arg, "chunks")]
+    tuple_args = [arg for arg in args if _is_chunk_tuple(arg)]
+    if not tuple_args and not chunked_args:
+        raise RuntimeError("Cannot determine desired output chunksize for cached function.")
```
Comment (mraspaud) [inline satpy/modifiers/angles.py:219]: Is this line tested?

File: satpy/utils.py [hunk 2]
```diff
@@ -40,6 +40,10 @@
 TRACE_LEVEL = 5
 
 
+class PerformanceWarning(Warning):
+    """Warning raised when there is a possible performance impact."""
```
Comment (mraspaud) [inline satpy/utils.py:44]: There is a similar warning class in dask, should we not use that instead?
https://docs.dask.org/en/latest/generated/dask.array.core.PerformanceWarning.html

File: satpy/tests/test_modifiers.py [hunk 3]
```diff
@@ -593,8 +599,18 @@ def test_get_angles_satpos_preference(self, forced_preference):
...
-    @pytest.mark.parametrize("input_func", [_get_angle_test_data, _get_stacked_angle_test_data])
-    def test_cache_get_angles(self, input_func, input2_func, exp_equal_sun, exp_num_zarr, force_bad_glob, tmpdir):
+    @pytest.mark.parametrize(
+        ("input_func", "num_normalized_chunks"),
+        [
+            (_get_angle_test_data, 9),
+            (_get_stacked_angle_test_data, 3),
+            (_get_angle_test_data_odd_chunks, 9),
+        ])
+    def test_cache_get_angles(
```
Comment (mraspaud) [inline satpy/tests/test_modifiers.py:609]: This function is starting to get long, can it be split somehow?

File: satpy/modifiers/angles.py [hunk 4]
```diff
@@ -231,11 +269,52 @@ def _sanitize_observer_look_args(*args):
...
+        else:
+            new_args.append(arg)
+    return new_args
+
+
+def _is_chunk_tuple(some_obj: Any) -> bool:
+    if not isinstance(some_obj, tuple):
+        return False
+    if not all(isinstance(sub_obj, tuple) for sub_obj in some_obj):
+        return False
```
Comment (mraspaud) [inline satpy/modifiers/angles.py:295]: Is this line tested?

File: satpy/modifiers/angles.py [hunk 5]
```diff
@@ -231,11 +269,52 @@ def _sanitize_observer_look_args(*args):
         elif isinstance(arg, (float, np.float64, np.float32)):
             # round floating point numbers to nearest tenth
             new_args.append(round(arg, 1))
+        elif _is_chunk_tuple(arg) and _chunks_are_irregular(arg):
+            new_chunks = _regular_chunks_from_irregular_chunks(arg)
+            new_args.append(new_chunks)
```
Comment (mraspaud) [inline satpy/modifiers/angles.py:274]: I suppose it's not possible to use the more pythonic try/except instead of checking the regularity of the chunks preemptivly?

**Review by mraspaud — COMMENTED** [review]
I'm surprised this functionality is not provided by dask or xarray, or even zarr itself. Would it make more sense to implement it there?
Otherwise it looks good, but I still do have a few comments :)

File: satpy/utils.py (hunk 2 above)
Comment (djhoese) [reply satpy/utils.py:44]: Pyresample also has a copy of the PerformanceWarning already in it. I'm not sure how I feel about depending on dask for a warning like this. I think in the pyresample case it was a PerformanceWarning about giving a numpy array to a dask resampler or something like that so the PerformanceWarning wasn't specifically about chunking like dask's `PerformanceWarning` says in the docstring.

In this case, it does match that description, or at least matches more than the pyresample case did.

Overall I'd prefer not to, but if you feel strongly I can do it. We may want a PerformanceWarning for other things in Satpy beyond chunking.

File: satpy/modifiers/angles.py (hunk 5 above)
Comment (djhoese) [reply satpy/modifiers/angles.py:274]: I'd prefer not to given the number of errors/exceptions that could arise from calling in to `to_zarr` that may or may not be related to irregular chunking.

File: satpy/tests/test_modifiers.py (hunk 3 above)
Comment (djhoese) [reply satpy/tests/test_modifiers.py:609]: CodeScene was fine with it :wink: 

I'll see what I can do.

File: satpy/modifiers/angles.py (hunk 1 above)
Comment (djhoese) [reply satpy/modifiers/angles.py:219]: No and it isn't actually possible to hit it with the current implemented methods because `chunks` is always provided. I could update this function to be less future-proof and not look for dask arrays and only look for chunk tuples.

File: satpy/modifiers/angles.py (hunk 4 above)
Comment (djhoese) [reply satpy/modifiers/angles.py:295]: No and it isn't possible right now because all the calling functions of the cached functions provide cache tuples and no other tuples. I could make a fake function to cache and test these cases...let me try that.

File: satpy/utils.py (hunk 2 above)
Comment (mraspaud) [reply satpy/utils.py:44]: Ok, just asking. We can keep this.

File: satpy/modifiers/angles.py (hunk 4 above)
Comment (mraspaud) [reply satpy/modifiers/angles.py:295]: if this code is unreachable anyway, we could just remove it then?

File: satpy/modifiers/angles.py (hunk 5 above)
Comment (mraspaud) [reply satpy/modifiers/angles.py:274]: ok

**Comment by djhoese** [comment]
This should be fixed now. Code simplified (I think) and additional tests were added to test the zarr caching edge cases that the builtin functions weren't covering.

**Review by mraspaud — APPROVED** [review]
LGTM
